botorch/models/transforms/input.py

```diff
@@ -1625,6 +1625,153 @@ def _expanded_perturbations(self, X: Tensor) -> Tensor:
+        self.transform_on_eval = transform_on_eval
+        self.transform_on_fantasize = transform_on_fantasize
+
+        self.encoders = encoders
+        self.categorical_features = categorical_features
+
+        if len(self.categorical_features) == 0:
+            raise ValueError(
+                "At least one categorical feature for encoding must be provided."
+            )
+
+        if len(self.categorical_features) > dim:
+            raise ValueError(
+                "The number of categorical features exceeds the provided dimension."
+            )
+
+        for idx, card in self.categorical_features.items():
+            if card == 1:
+                raise ValueError(
+                    f"Categorical feature at index {idx} has cardinality 1, "
+                    f"which is not allowed."
+                )
+
+        # check that the encoders match the categorical features
+        if set(self.encoders.keys()) != set(self.categorical_features.keys()):
+            raise ValueError(
+                "The keys of `encoders` must match the keys of `categorical_features`."
+            )
+
+        self.ordinal_idx = list(
+            self.categorical_features.keys()
+        )  # indices of categorical features before encoding
+
+        self.numerical_idx = list(
+            set(range(dim)) - set(self.ordinal_idx)
+        )  # indices of numerical features before encoding
+
+        self.new_numerical_idx = []  # indices of numerical features after encoding
+        self.encoded_idx = []  # indices of categorical features after encoding
+
+        offset = 0
+        for idx in range(dim):
+            if idx in self.numerical_idx:
+                self.new_numerical_idx.append(idx + offset)
+            else:
+                card = self.categorical_features[idx]
+                self.encoded_idx.append(
+                    np.arange(
+                        idx + offset, idx + offset + card
+                    ).tolist()  # indices of categorical features after encoding
+                )
+                offset += card - 1  # adjust offset for next categorical feature
+
+    def transform(self, X: Tensor) -> Tensor:
+        r"""Transform the categorical inputs into a vector representation.
+
+        Args:
+            X: A `batch_shape x n x d`-dim tensor of inputs.
+
+        Returns:
+            A `batch_shape x n x d'`-dim tensor of where the integer encoded
+            categoricals are transformed to a vector representation.
+        """
+        s = list(X.shape)
+        s[-1] = len(self.numerical_idx) + len(np.concatenate(self.encoded_idx))
+        X_encoded = torch.zeros(size=s).to(X)
+        X_encoded[..., self.new_numerical_idx] = X[..., self.numerical_idx]
+        for i, idx in enumerate(self.categorical_features.keys()):
+            X_encoded[..., self.encoded_idx[i]] = self.encoders[idx](
+                X[..., idx].long(),
+            ).to(X_encoded)
```

Comment:
slight readability tweak
```suggestion
            encoded_val = self.encoders[idx](X[..., idx].long()).to(X_encoded)
            X_encoded[..., self.encoded_idx[i]] = encoded_val
```